smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelAssembler.java

```diff
@@ -544,16 +545,41 @@ public ModelAssembler validationEventListener(Consumer<ValidationEvent> eventLis
             traitFactory = LazyTraitFactoryHolder.INSTANCE;
         }
 
+        List<ModelFile> files = createModelFiles();
         // Create "model files" for the prelude, manually added shapes, imports, etc.
-        CompositeModelFile composite = new CompositeModelFile(traitFactory, createModelFiles());
+        CompositeModelFile composite = new CompositeModelFile(traitFactory, files);
 
         try {
             TraitContainer traits = composite.resolveShapes(composite.shapeIds(), composite::getShapeType);
             Model model = Model.builder()
                     .metadata(composite.metadata())
                     .addShapes(composite.createShapes(traits).getCreatedShapes())
                     .build();
-            return validate(model, traits, composite.events());
+
+            List<ValidationEvent> compositeEvents = composite.events();
+
+            // Always perform trait validation before 1.0 -> 2.0 transforms.
+            validateTraits(model.getShapeIds(), traits, compositeEvents);
+            traits = null;
+
+            // If ERROR validation events occur while loading, then performing more
+            // granular semantic validation will only obscure the root cause of errors.
+            if (disableValidation || LoaderUtils.containsErrorEvents(compositeEvents)) {
```

Comment:
Is `disableValidation` skipping doing an upgrade because we won't know if we broke the model vs received a broken one? Is this worth calling out in the docs for `disableValidation`?